framework/src/modules/interoperability/utils.ts

```diff
@@ -106,3 +146,295 @@ export const computeValidatorsHash = (
+		if (
+			decodedCertificate.blockID.equals(EMPTY_BYTES) ||
+			decodedCertificate.stateRoot.equals(EMPTY_BYTES) ||
+			decodedCertificate.validatorsHash.equals(EMPTY_BYTES) ||
+			decodedCertificate.timestamp === 0
+		) {
+			return {
+				status: VerifyStatus.FAIL,
+				error: new Error('Certificate is missing required values.'),
+			};
+		}
+
+		// Last certificate height should be less than new certificate height
+		if (decodedCertificate.height <= partnerChainAccount.lastCertificate.height) {
+			return {
+				status: VerifyStatus.FAIL,
+				error: new Error('Certificate height should be greater than last certificate height.'),
+			};
+		}
+	}
+
+	return {
+		status: VerifyStatus.OK,
+	};
+};
+
+export const checkActiveValidatorsUpdate = (
+	txParams: CrossChainUpdateTransactionParams,
+): VerificationResult => {
+	if (
+		txParams.activeValidatorsUpdate.length !== 0 ||
+		txParams.newCertificateThreshold > BigInt(0)
+	) {
+		// Non-empty certificate
+		if (txParams.certificate.equals(EMPTY_BYTES)) {
+			return {
+				status: VerifyStatus.FAIL,
+				error: new Error(
+					'Certificate cannot be empty when activeValidatorsUpdate is non-empty or newCertificateThreshold >0.',
+				),
+			};
+		}
+
+		// params.activeValidatorsUpdate has the correct format
+		for (let i = 0; i < txParams.activeValidatorsUpdate.length; i += 1) {
+			const currentValidator = txParams.activeValidatorsUpdate[i];
+			const nextValidator = txParams.activeValidatorsUpdate[i + 1];
+			if (currentValidator.blsKey.byteLength !== VALID_BLS_KEY_LENGTH) {
+				return {
+					status: VerifyStatus.FAIL,
+					error: new Error(`BlsKey length should be equal to ${VALID_BLS_KEY_LENGTH}.`),
+				};
+			}
+			if (currentValidator.blsKey.compare(nextValidator.blsKey) > -1) {
+				return {
+					status: VerifyStatus.FAIL,
+					error: new Error('Validators blsKeys must be unique and lexicographically ordered'),
+				};
+			}
+		}
+	}
+
+	return {
+		status: VerifyStatus.OK,
+	};
+};
+
+export const checkInboxUpdateValidity = (
+	txParams: CrossChainUpdateTransactionParams,
+	partnerChannelData: ChannelData,
+): VerificationResult => {
+	const partnerChainIDBuffer = getIDAsKeyForStore(txParams.sendingChainID);
+	const decodedCertificate = codec.decode<Certificate>(certificateSchema, txParams.certificate);
+	const { crossChainMessages, messageWitness, outboxRootWitness } = txParams.inboxUpdate;
+	const ccmHashes = crossChainMessages.map(ccm => hash(ccm));
+
+	let newInboxRoot;
+	let newInboxAppendPath = partnerChannelData.inbox.appendPath;
+	let newInboxSize = partnerChannelData.inbox.size;
+	for (const ccm of ccmHashes) {
+		const { appendPath, size } = regularMerkleTree.calculateMerkleRoot({
+			value: ccm,
+			appendPath: newInboxAppendPath,
+			size: newInboxSize,
+		});
+		newInboxAppendPath = appendPath;
+		newInboxSize = size;
+	}
+	// If inboxUpdate contains a non-empty messageWitness, then update newInboxRoot to the output
```

Comment:
from the LIP, condition seems like
if cerficate is not empty && inboxUpdate is not empty
  if inboxUpdate contains non-empty message witness, update the root
check the root
elseif cerificate is empty && inboxUPdate is not empty
  if inboxUpdate contains non-empty message witness, update the root
check the root

so it could be like
```
if imboxUpdate is empty, return
if certrificate is nonempty
  if messageWitness is non-empty, update root
else
  if messageWitness is non-empty, update root

check the root
```
something like this?